skfuzzy/defuzzify/defuzz.py

```diff
@@ -58,7 +58,43 @@ def centroid(x, mfx):
     skfuzzy.defuzzify.defuzz, skfuzzy.defuzzify.dcentroid
 
     """
-    return (x * mfx).sum() / np.fmax(mfx.sum(),
+
+    '''
+    As we suppose linearity between each pair of points of x, we can calculate the exact area of the figure
+    (a triangle or a rectangle).
+    '''
+    sum_moment_area = 0.0
+    sum_area = 0.0
+    #If the membership function is a singleton fuzzy set:
+    if len(x)==1:
+        return x[0]*mfx[0]/np.fmax(mfx[0],
+                                     np.finfo(float).eps).astype(float)
+    #else return the sum of moment*area/sum of area
+    for i in range(1,len(x)):
```

Comment:
I'm a little concerned about performance here. At first I thought this loop would be well served by vectorization. However, the conditionals make that a challenge.

I'm fine leaving this approach for now, but likely this would reap huge performance gains from Cython.